flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcFileWriter.java

```diff
@@ -0,0 +1,269 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.orc.OrcBatchReader.schemaToTypeInfo;
+
+/**
+ * A {@link Writer} that writes the bucket files as Hadoop {@link OrcFile}.
+ *
+ * @param <T> The type of the elements that are being written by the sink.
+ */
+public class OrcFileWriter<T extends Row> extends StreamWriterBase<T> {
+
+	private static final long serialVersionUID = 3L;
+
+	/**
+	 * The description of the types in an ORC file.
+	 */
+	private TypeDescription schema;
+
+	/**
+	 * The schema of an ORC file.
+	 */
+	private String metaSchema;
+
+	/**
+	 * A row batch that will be written to the ORC file.
+	 */
+	private VectorizedRowBatch rowBatch;
+
+	/**
+	 * The writer that fill the records into the batch.
+	 */
+	private OrcBatchWriter orcBatchWriter;
+
+	private transient org.apache.orc.Writer writer;
+
+	private CompressionKind compressionKind;
+
+	/**
+	 * The number of rows that currently being written.
+	 */
+	private long writedRowSize;
+
+	/**
+	 * Creates a new {@code OrcFileWriter} that writes orc files without compression.
+	 *
+	 * @param metaSchema The orc schema.
+	 */
+	public OrcFileWriter(String metaSchema) {
+		this(metaSchema, CompressionKind.NONE);
+	}
+
+	/**
+	 * Create a new {@code OrcFileWriter} that writes orc file with the gaven
+	 * schema and compression kind.
+	 *
+	 * @param metaSchema      The schema of an orc file.
+	 * @param compressionKind The compression kind to use.
+	 */
+	public OrcFileWriter(String metaSchema, CompressionKind compressionKind) {
+		this.metaSchema = metaSchema;
+		this.schema = TypeDescription.fromString(metaSchema);
+		this.compressionKind = compressionKind;
+	}
+
+	@Override
+	public void open(FileSystem fs, Path path) throws IOException {
+		writer = OrcFile.createWriter(path, OrcFile.writerOptions(fs.getConf()).setSchema(schema).compress(compressionKind));
+		rowBatch = schema.createRowBatch();
+		orcBatchWriter = new OrcBatchWriter(Arrays.asList(orcSchemaToTableSchema(schema).getTypes()));
+	}
+
+	private TableSchema orcSchemaToTableSchema(TypeDescription orcSchema) {
+		List<String> fieldNames = orcSchema.getFieldNames();
+		List<TypeDescription> typeDescriptions = orcSchema.getChildren();
+		List<TypeInformation> typeInformations = new ArrayList<>();
+
+		typeDescriptions.forEach(typeDescription -> {
+			typeInformations.add(schemaToTypeInfo(typeDescription));
+		});
+
+		return new TableSchema(
+			fieldNames.toArray(new String[fieldNames.size()]),
+			typeInformations.toArray(new TypeInformation[typeInformations.size()]));
+	}
+
+	@Override
+	public void write(T element) throws IOException {
+		Boolean isFill = orcBatchWriter.fill(rowBatch, element);
+		if (!isFill) {
+			writer.addRowBatch(rowBatch);
+			writedRowSize += writedRowSize + rowBatch.size;
+			rowBatch.reset();
+		}
+	}
+
+	@Override
+	public long flush() throws IOException {
+		writer.addRowBatch(rowBatch);
```

Comment:
This call does not guarantee that the data would be persisted. The data might still stay in memory and getting lost in case of fail over.

Another problem is that even the data is getting persisted, this orc file might be unreadable since there is no proper File footer there. The correct flushing behaviour should make sure that the data has been persisted and a proper file footer has been created.

Reply:
You are very correct!